react-common/components/controls/TeachingBubble.tsx

```diff
@@ -0,0 +1,291 @@
+        window.addEventListener("resize", positionBubbleAndCutout);
+        return () => {
+            window.removeEventListener("resize", positionBubbleAndCutout);
+        }
+    }, [stepNumber]);
+
+    useEffect(() => {
+        if (!hasSteps) {
+            const footer = document.querySelector(".teaching-bubble-footer") as HTMLElement;
+            footer.style.flexDirection = "row-reverse";
+        }
+    }, []);
+
+    const positionBubbleAndCutout = () => {
+        bubble = document.getElementById(id);
+        bubbleArrow = document.querySelector(".teaching-bubble-arrow") as HTMLElement;
+        bubbleArrow.style.border = "none";
+        bubbleBounds = bubble.getBoundingClientRect();
+        // To Do: check that targetContent.targetQuery is a valid selector
+        targetElement = document.querySelector(targetContent.targetQuery) as HTMLElement;
+        if (!targetElement) {
+            // display bubble in center of screen
+            updatePosition(bubble, `${(window.innerHeight - bubbleBounds.height) / 2}px`, `${(window.innerWidth - bubbleBounds.width) / 2}px`);
+            clearCutout();
+            return;
+        }
+        targetBounds = targetElement.getBoundingClientRect();
+        setPosition();
+        setCutout();
+    }
+
+    const setCutout = () => {
+        const cutout = document.querySelector(".teaching-bubble-cutout") as HTMLElement;
+        let cutoutTop = targetBounds.top;
+        let cutoutLeft = targetBounds.left;
+        let cutoutWidth = targetBounds.width;
+        let cutoutHeight = targetBounds.height;
+        // make cutout bigger if no padding and not centered
+        if (targetContent.location !== Location.Center) {
+            const paddingTop = parseFloat(window.getComputedStyle(targetElement).paddingTop);
+            const paddingRight = parseFloat(window.getComputedStyle(targetElement).paddingRight);
+            const paddingBottom = parseFloat(window.getComputedStyle(targetElement).paddingBottom);
+            const paddingLeft = parseFloat(window.getComputedStyle(targetElement).paddingLeft);
+            if (paddingTop < margin) {
+                cutoutTop -= margin + paddingTop;
+                cutoutHeight += margin - paddingTop;
+            }
+            if (paddingBottom < margin) {
+                cutoutHeight += margin - paddingBottom;
+            }
+            if (paddingLeft < margin) {
+                cutoutLeft -= margin + paddingLeft;
+                cutoutWidth += margin - paddingLeft;
+            }
+            if (paddingRight < margin) {
+                cutoutWidth += margin - paddingRight;
+            }
+        }
+        cutout.style.top = `${cutoutTop}px`;
+        cutout.style.left = `${cutoutLeft}px`;
+        cutout.style.width = `${cutoutWidth}px`;
+        cutout.style.height = `${cutoutHeight}px`;
+
+        if (activeTarget) {
+            cutout.style.pointerEvents = "none";
+        }
+    }
+
+    const clearCutout = () => {
```

Comment:
Good suggestion with making this based on state. Instead of adding a class, I'll just make that div conditional depending on the state.

Reply:
Actually, the div needs to always exist, so I'll do it with classes.

Reply:
Sadly, it won't work with classes because when the styles are set in code, they are set at the element level, which overrides anything at the class level. So, it looks like I will need to leave it as is.

Reply:
Ah, yeah, definitely a frustrating thing with CSS. Sounds good!